### PR TITLE
#1548 Update logic for ISL with "Under maintenance" flag.

### PR DESCRIPTION
**Pull request by dpoltavets**
Part of issue: #1548.

Depends on: #1742.

---
This change is [Reviewable](https://reviewable.io/reviews/telstra/open-kilda/1754)

### CODE REVIEW

File: services/wfm/src/main/java/org/openkilda/wfm/topology/event/service/IslService.java [hunk 1]
```diff
@@ -58,10 +58,14 @@ public IslService(TransactionManager transactionManager, RepositoryFactory repos
...
-                isl.getDestSwitch().getSwitchId(), isl.getDestPort(), isl.getStatus());
-        sender.sendRerouteInactiveFlowsMessage(reason);
+        Optional<Isl> foundIsl = islRepository.findByEndpoints(isl.getSrcSwitch().getSwitchId(), isl.getSrcPort(),
+                isl.getDestSwitch().getSwitchId(), isl.getDestPort());
+        if (foundIsl.isPresent() && !foundIsl.get().isUnderMaintenance()) {
+            String reason = String.format("Create or update ISL: %s_%d-%s_%d. ISL status: %s",
+                    isl.getSrcSwitch().getSwitchId(), isl.getSrcPort(),
+                    isl.getDestSwitch().getSwitchId(), isl.getDestPort(), isl.getStatus());
+            sender.sendRerouteInactiveFlowsMessage(reason);
+        }
```
Comment (timofei-durakov) [inline services/wfm/src/main/java/org/openkilda/wfm/topology/event/service/IslService.java]: tbh I don't like this side effect of create or update method

File: services/wfm/src/main/java/org/openkilda/wfm/topology/reroute/service/RerouteService.java [hunk 2]
```diff
@@ -51,4 +52,20 @@ public RerouteService(RepositoryFactory repositoryFactory) {
...
+     *
+     * @param srcSwitchId source switch id.
+     * @param srcPort     source port.
+     * @param dstSwitchId destination switch id.
+     * @param dstPort     destination port.
+     */
+    public Collection<String> getAllFlowIdsForIsl(SwitchId srcSwitchId, int srcPort,
+                                               SwitchId dstSwitchId, int dstPort) {
+        log.info("Get inactive flow ids");
+        return flowRepository.findAllFlowPairsWithSegment(srcSwitchId, srcPort, dstSwitchId, dstPort).stream()
```
Comment (timofei-durakov) [inline services/wfm/src/main/java/org/openkilda/wfm/topology/reroute/service/RerouteService.java]: why not to do search over flow segments, not flows?

File: services/wfm/src/test/java/org/openkilda/wfm/topology/nbworker/bolts/LinkOperationsBoltTest.java [hunk 3]
```diff
@@ -73,7 +73,7 @@ public void shouldCreateLinkProps() throws AbstractException {
         switchRepository.createOrUpdate(Switch.builder().switchId(SWITCH_ID_1).build());
         switchRepository.createOrUpdate(Switch.builder().switchId(SWITCH_ID_2).build());
 
-        LinkOperationsBolt bolt = new LinkOperationsBolt(persistenceManager);
+        LinkOperationsBolt bolt = new LinkOperationsBolt(persistenceManager, 10000);
```
Comment (timofei-durakov) [inline services/wfm/src/test/java/org/openkilda/wfm/topology/nbworker/bolts/LinkOperationsBoltTest.java]: constant please

File: services/src/messaging/src/main/java/org/openkilda/messaging/command/reroute/RerouteAllFlowsForIsl.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.EqualsAndHashCode;
+import lombok.NonNull;
+import lombok.Value;
+
+@Value
+@EqualsAndHashCode(callSuper = false)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class RerouteAllFlowsForIsl extends RerouteFlows {
```
Comment (timofei-durakov) [inline services/src/messaging/src/main/java/org/openkilda/messaging/command/reroute/RerouteAllFlowsForIsl.java]: do we really need new RerouteFlows subtype here? why can't we utilize reroute affected flows instead?

Comment (dpoltavets) [reply services/src/messaging/src/main/java/org/openkilda/messaging/command/reroute/RerouteAllFlowsForIsl.java]: We should not reroute affected flows in this case. We have to reroute all flows which go through a particular ISL.

File: services/wfm/src/main/java/org/openkilda/wfm/topology/reroute/service/RerouteService.java (hunk 2 above)
Comment (dpoltavets) [reply services/wfm/src/main/java/org/openkilda/wfm/topology/reroute/service/RerouteService.java]: I just used an existing method to search for a flow ids by ISL. And by code, we are searching flow ids over flow segments.

File: services/wfm/src/main/java/org/openkilda/wfm/topology/event/service/IslService.java [hunk 5]
```diff
@@ -58,10 +58,14 @@ public IslService(TransactionManager transactionManager, RepositoryFactory repos
     public void createOrUpdateIsl(Isl isl, Sender sender) {
         createOrUpdateIsl(isl);
 
-        String reason = String.format("Create or update ISL: %s_%d-%s_%d. ISL status: %s",
-                isl.getSrcSwitch().getSwitchId(), isl.getSrcPort(),
-                isl.getDestSwitch().getSwitchId(), isl.getDestPort(), isl.getStatus());
-        sender.sendRerouteInactiveFlowsMessage(reason);
+        Optional<Isl> foundIsl = islRepository.findByEndpoints(isl.getSrcSwitch().getSwitchId(), isl.getSrcPort(),
+                isl.getDestSwitch().getSwitchId(), isl.getDestPort());
+        if (foundIsl.isPresent() && !foundIsl.get().isUnderMaintenance()) {
```
Comment (sergii-iakovenko) [inline services/wfm/src/main/java/org/openkilda/wfm/topology/event/service/IslService.java]: Why shouldn't we try to reroute inactive flows if previously unknown link appeared?

File: services/src/messaging/src/main/java/org/openkilda/messaging/command/reroute/RerouteAllFlowsForIsl.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    @JsonCreator
+    public RerouteAllFlowsForIsl(@NonNull @JsonProperty("source") PathNode source,
+                                 @NonNull @JsonProperty("destination") PathNode destination,
+                                 @NonNull @JsonProperty("reason") String reason) {
+        super(reason);
+        this.source = source;
+        this.destination = destination;
+    }
+
+    public PathNode getSource() {
```
Comment (sergii-iakovenko) [inline services/src/messaging/src/main/java/org/openkilda/messaging/command/reroute/RerouteAllFlowsForIsl.java]: Excessive as @Value is used and lombok generates getters.

File: services/wfm/src/main/java/org/openkilda/wfm/topology/reroute/bolts/RerouteBolt.java [hunk 7]
```diff
@@ -98,6 +99,17 @@ protected void handleInput(Tuple tuple) {
             emitRerouteCommands(tuple, inactiveFlows, message.getCorrelationId(),
                     rerouteInactiveFlows.getReason());
 
+        } else if (commandData instanceof RerouteAllFlowsForIsl) {
+            RerouteAllFlowsForIsl rerouteAllFlowsForIsl = (RerouteAllFlowsForIsl) commandData;
+            PathNode source = rerouteAllFlowsForIsl.getSource();
+            PathNode destination = rerouteAllFlowsForIsl.getDestination();
+            Collection<String> inactiveFlows =
```
Comment (sergii-iakovenko) [inline services/wfm/src/main/java/org/openkilda/wfm/topology/reroute/bolts/RerouteBolt.java]: Why are they inactive?

File: services/wfm/src/main/java/org/openkilda/wfm/topology/reroute/bolts/RerouteBolt.java [hunk 8]
```diff
@@ -98,6 +99,17 @@ protected void handleInput(Tuple tuple) {
...
+                    rerouteService.getAllFlowIdsForIsl(source.getSwitchId(), source.getPortNo(),
+                                                       destination.getSwitchId(), destination.getPortNo());
+
+            emitRerouteCommands(tuple, inactiveFlows, message.getCorrelationId(),
```
Comment (sergii-iakovenko) [inline services/wfm/src/main/java/org/openkilda/wfm/topology/reroute/bolts/RerouteBolt.java]: The only difference I see between RerouteAffectedFlows and RerouteAllFlowsForIsl is that the latter causes reroute for inactive flows too. Am I wrong?

File: services/wfm/src/main/java/org/openkilda/wfm/topology/event/service/IslService.java (hunk 5 above)
Comment (dpoltavets) [reply services/wfm/src/main/java/org/openkilda/wfm/topology/event/service/IslService.java]: `UP. do nothing (even not issue reroute for flows in DOWN state)` #1548

File: services/wfm/src/main/java/org/openkilda/wfm/topology/reroute/bolts/RerouteBolt.java (hunk 8 above)
Comment (dpoltavets) [reply services/wfm/src/main/java/org/openkilda/wfm/topology/reroute/bolts/RerouteBolt.java]: No, you are right.

**Comment by dpoltavets** [comment]
It should be in Nbworker topology for now and will be refactored after #1834 and implementation of #1824.

**Pull request merged**